### PR TITLE
don't erase existing body classnames

### PR DESCRIPTION
**Pull request by mikhuang**
I was having a problem with [reactstrap](http://reactstrap.github.io) where the `modal-open` class name it set on &lt;body&gt; was erased. This pull request tracks which classes have been added by ReactBodyClassname and preserves other classes set on &lt;body&gt;.

### CODE REVIEW

File: test/browser.spec.js [hunk 1]
```diff
@@ -19,7 +19,15 @@ describe('BodyClassName (in a browser)', function () {
     expect(global.document.body.className).to.equal(className);
   });
 
+  it('does not erase existing body class names', function () {
+    global.document.body.className = 'testing'
+    var className = 'hello world';
+    var Component = enzyme.mount(React.createElement(BodyClassName, {className: className}));
+    expect(global.document.body.className).to.equal('testing hello world');
+  });
```
Comment (iest) [inline test/browser.spec.js:27]: Could you add a test that it doesn't duplicate classnames that already exist on the body _and_ have been added by this lib? I'm thinking about SSR and want to make sure the base is covered.

File: index.js [hunk 2]
```diff
@@ -4,6 +4,12 @@ var React = require('react');
 var withSideEffect = require('react-side-effect');
 var PropTypes = require('prop-types');
 
+let classNameCache = []
```
Comment (iest) [inline index.js:7]: I think the cache should be a static property on `BodyClassName` itself — it might be useful to expose it for people to use.

File: index.js [hunk 3]
```diff
@@ -13,7 +19,26 @@ function reducePropsToState(propsList) {
...
+  for (let i = 0; i < classNameCache.length; i++) {
+    const idx = currentClassNames.indexOf(classNameCache[i]);
+    if (idx !== -1) {
+      currentClassNames.splice(idx, 1);
+    }
+  }
+  
+  // append all new ones
+  const newClassNames = splitClassName(stringClassNames);
+  currentClassNames = currentClassNames.concat(newClassNames);
```
Comment (iest) [inline index.js:35]: I'm not a huge fan of the imperative-mutation approach. A functional approach would be better imho.
Maybe something like:

```js
// Lets say we already added some
BodyClassName.addedClassNames = ['foo', 'bar', 'baz'];

var currentClassNames = splitClassName(document.body.className).filter(function(
  className
) {
  return cache.indexOf(className) !== -1;
});

var newClassNames = currentClassNames
  .concat(splitClassName(stringClassNames))
  .join(' ')
  .trim();
  
BodyClassName.addedClassNames = newClassNames;
document.body.className = newClassNames;
```

**Review by iest — CHANGES_REQUESTED** [review]
Firstly, thanks so much for opening a PR!

Secondly, apologies for the slow response. I wasn't sure if the library should support your use case — never too sure about duplicating state, and wasn't convinced it was necessary.

However, your implementation seems to work well and I _don't think_ it has any negative effects (or any potential memory leak issues for SSR which I was initially worried about).

Just a couple changes to make the code easier to understand and it's good to go ✨

**Comment by iest** [comment]
@mikhuang any progress on these changes? I'll probably close the PR soon due to inactivity if not.

Thanks!

**Comment by JiiB** [comment]
@mikhuang This is exactly what I need. @iest I'm going to make a PR with mikhuang's changes and your feedback request.

**Comment by iest** [comment]
Closing in favour of #8
